Add error handling, output & test harness

wadells opened the pull request:
This PR addresses some previous feedback (#7 and #8) and cleans up a bunch of 'it works but is fragile'.  Now we have nice error messages, and teardown even if the test fails.

The changes in `src/grav/test.py` look big because indentation changes but they're actually small (`assert` changes, unittest integration, and some comments)

Testing done:
```
# without docker running

(venv) vagrant@ubuntu-bionic:/vagrant$ python3 src/grav/main.py
test_start_stop_logs (grav.test.ContainerStartStopLogTest) ... ERROR

======================================================================
ERROR: test_start_stop_logs (grav.test.ContainerStartStopLogTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/vagrant/src/grav/client.py", line 99, in __init__
    self.conn.connect()
  File "/vagrant/src/grav/client.py", line 78, in connect
    sock.connect(self.socket_path)
ConnectionRefusedError: [Errno 111] Connection refused

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/vagrant/src/grav/test.py", line 47, in setUp
    client = DockerClient("/var/run/docker.sock")
  File "/vagrant/src/grav/client.py", line 102, in __init__
    raise DockerClientError(msg) from e
grav.client.DockerClientError: Unable to connect to '/var/run/docker.sock'. Is the docker daemon running?

----------------------------------------------------------------------
Ran 1 test in 0.007s

FAILED (errors=1)


# killing docker part way through setup

(venv) vagrant@ubuntu-bionic:/vagrant$ python3 src/grav/main.py
test_start_stop_logs (grav.test.ContainerStartStopLogTest) ... ERROR

======================================================================
ERROR: test_start_stop_logs (grav.test.ContainerStartStopLogTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/vagrant/src/grav/client.py", line 114, in request
    self.conn.request(verb, path, body=body, headers=headers)
  File "/usr/lib/python3.6/http/client.py", line 1254, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.6/http/client.py", line 1300, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.6/http/client.py", line 1249, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.6/http/client.py", line 1036, in _send_output
    self.send(msg)
  File "/usr/lib/python3.6/http/client.py", line 996, in send
    self.sock.sendall(data)
BrokenPipeError: [Errno 32] Broken pipe

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/vagrant/src/grav/test.py", line 68, in setUp
    resp = client.post("/containers/create", body=data, headers=headers)
  File "/vagrant/src/grav/client.py", line 108, in post
    return self.request("POST", path, body=body, headers=headers)
  File "/vagrant/src/grav/client.py", line 119, in request
    raise DockerClientError(msg) from e
grav.client.DockerClientError: Failure during 'POST /containers/create request to '/var/run/docker.sock'

# a forced failure:

(venv) vagrant@ubuntu-bionic:/vagrant$ python3 src/grav/main.py
test_start_stop_logs (grav.test.ContainerStartStopLogTest) ... FAIL

======================================================================
FAIL: test_start_stop_logs (grav.test.ContainerStartStopLogTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/vagrant/src/grav/test.py", line 89, in test_start_stop_logs
    self.assertStatus(resp, 400)
  File "/vagrant/src/grav/harness.py", line 29, in assertStatus
    raise self.failureException(e) from None
AssertionError: POST /containers/5b8c4dfc5d049336a244d61507ed997cb7f4ebb6132a5b39fab3bd6ced4b514a/start returned http status 204 instead of expected 400.
Body:
b''

----------------------------------------------------------------------
Ran 1 test in 3.542s

FAILED (failures=1)
(venv) vagrant@ubuntu-bionic:/vagrant$ docker container ls -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
```
```

wadells committed:
Bind raw test functions to unittest.TestCase.

Instead of executing the test lifecycle as a simple script, I now use
the unittest harness to have more robust logging and error handling
(e.g. teardown will be run even if the test has an error).

I've also added a main file and function, with the intent of having that
be the entrypoint (as defined in setup.py) to this code.

Addresses:
    https://github.com/javins/gravitational-audition/issues/8
and
    https://github.com/javins/gravitational-audition/pull/5#discussion_r335759224

wadells committed:
Refactor the functions in test.py to be pure unittest methods.

In order to simplify the codebase and more importantly:
To allow use of unittest style `assert*` methods, the test source
needs access to the unittest test case (typically done via `self`).

I've chosen to remove the original function and only support running
this via unittest for simplicity's sake.

wadells committed:
Add a note about regexing images.

As this was a question in CR, it is clearly suspicious enough to leave a
note to future maintainers.

wadells committed:
Add an editorial rant about unittest.

I wasn't just gonna let this antiquity get a free endoresment.

wadells committed:
Add .body property to FriendlyHttpResponse.

In order to do logging well, we want to log response bodies.  However,
if a test has already called .read() on the response, the body is not
available to logging without passing it.

This is a bit of a hack to cache the body in the response object, so
that it can be accessed multiple times.  I believe the requests
response object does something like this.

wadells committed:
Remove unnecessary positional arg in front of *args.

We scatter everything anyhow... no need to call one param out in
particular, to simply pass it to super before the scattered args.

wadells committed:
Record the request verb and path on the response object.

This is a lazy backpointer from response -> request for logging, as
httplib doesn't have that by default.  This is a bit messy, and not
something I'd do in production.  But it meets my immeidate needs, and
greatly simplifies the test api.

wadells committed:
Improve output on http status mismatch.

As mentioned in:
    https://github.com/javins/gravitational-audition/issues/7
and
    https://github.com/javins/gravitational-audition/pull/5#discussion_r335759114

We want to give better messages when there is an assertion failure. This
patch handles all of the times http status codes could mismatch.  We
still want to deal with timeouts and socket errors though.

wadells committed:
Extract harness code into its own module.

To keep the test code clean we pull out more general harness stuff into
a seperate file.

wadells committed:
Add more useful error messages to the client.

Include the socket and any request being made when the client has
trouble.

wadells committed:
Add comment about danging resources.

wadells committed:
Add a comment about lack of cli parsing.

klizhentas commented on src/grav/client.py:
```diff
@@ -17,14 +17,44 @@
+    I really like requests .json() method on response objects.
+
+    Also, in vanilla HTTPResponse objects, once read() there is no sanctioned
+    API to get the body content a second time.  Hope you remembed it! This
```
`remembed` -> remembered?

wadells replied:
Addressed by 05315a7.

bernardjkim commented on src/grav/test.py:
```diff
@@ -1,116 +1,141 @@
+        stdout_lines = stdout.splitlines()
+        self.assertEqual(stdout_lines[0], "Hello Gravitational!")
+        self.assertEqual(stdout_lines[-1], "Terminated. Bye!")
+        self.assertEqual(stderr, "")
```
Better to assert that stderr is empty before checking stdout. Incase of err could stdout be empty?

wadells replied:
Good point. We'll get better error messages with this suggestion.

wadells replied:
Addressed by 9a3cd96.

r0mant commented on src/grav/client.py:
```diff
@@ -17,14 +17,44 @@
+    I really like requests .json() method on response objects.
+
+    Also, in vanilla HTTPResponse objects, once read() there is no sanctioned
+    API to get the body content a second time.  Hope you remembed it! This
```
Typo in "remembed".

wadells replied:
Addressed by 05315a7.

r0mant commented on src/grav/client.py:
```diff
@@ -17,14 +17,44 @@
+        # inheritance. -- wdella 2019-10
+        data = super(FriendlyHTTPResponse, self).read()
+        if self._body is None:
+            self._body = data
```
How would you do it if the response body was large? Not necessary for this project, I'm just curious.

wadells replied:
The right perf answer to this question largely depends on what we expect the size of the data to to be and the expected read pattern. In this case the expected size is < 1Mb and the read pattern is once, maybe up to 3 times in quick succession, after which the response object will fall out of scope and get GC'd.  I think the current implementation is fine for that.

To entertain your question a bit further, imagine that we had response size `O(1/2 * available system memory)` that needs to be read many times over a long period.  I'll say a response less than 1/2 the system memory can be streamed out of the socket and into a io buffer a la https://github.com/javins/gravitational-audition/pull/5/commits/07d71b367bcdc606cba30858dc7b741c52a24adf. Greater than that and we'll persist it to disk (/tmp probably) and open the file back when needed to be read again.  This could also play nicely with logging, as we'd already have response data on persistent storage.

I recognize with swap you could probably get 'free' volatile disk backed storage by jamming the whole thing into memory, but that feels ugly to my gut.

wadells replied:
Let me know if you have other questions.  Until then, marking resolved.

bernardjkim commented on src/grav/test.py:
```diff
@@ -1,116 +1,141 @@
+        delete_path = "/images/" + image_id
+        resp = client.request("DELETE", delete_path)
+        resp.read()  # TODO: have this handled by the client
+        # why isn't this 204 like the other api endpoints without a body?
```
https://docs.docker.com/engine/api/v1.40/#operation/ImageDelete
Looks like DELETE /images/{id} returns an array of deleted image ids.
Maybe worth verifying the returned ID matches up with the initial image ID.

wadells replied:
This is a great idea, but I'm going to leave it for the time being because it isn't as high priority to this project as getting the readme & last of the build work done.

Here is why: a typical response json looks like:

```
[
  {'Deleted': 'sha256:3ca1723f55b595b5646967b8b333be735cda2846c77b0f73fd606d8fb6f55f16'},
  {'Deleted': 'sha256:b208507827cb7c5b61fda4b1901a12a7d7d62ef5c81105af32a42de96d247b55'},
  {'Deleted': 'sha256:03901b4a2ea88eeaad62dbe59b072b28b6efa00491962b8741081c5df50c65e0'
]
```

There may be `Untagged`keys in those single item maps too, as well as other stuff?  I'd need to spend some more time reading the docs.  Given this format there isn't a concise known path I could pull out of the response (`json[0]['Deleted']` is a guess, but I wouldn't bank on the determinism of that). To do it right I would instead need some minor scanning logic to verify image deletion.

wadells replied:
#10 filed for this to be dealt with in future work.

wadells committed:
Fix typo.

Addresses
    https://github.com/javins/gravitational-audition/pull/9#discussion_r337122780
and
    https://github.com/javins/gravitational-audition/pull/9#discussion_r336746791

wadells committed:
Validate stdout after stderr.

If stderr is non-empty, we might expect the stdout stream to be
corrupted some how, as there has been an error.  Thus check stderr
first, so that it will show up in the test logs, and potentially be
useful to debugging.

Addresses https://github.com/javins/gravitational-audition/pull/9#discussion_r337121072

wadells committed:
Fix 201 -> 204 http status.

In eed158e5 I accidentally converted an expected 201 to and expected
204.  I'd typically just squash this commit into eed158e5, but I'm not
sure how Gravitational feels about rewriting history that is under
review in a PR. So... for now this remains as its own commit.

The test was broken without this, failing with:
    AssertionError: POST /containers/<id>/start returned http status 204 instead of expected 201.
